snuba/admin/static/query_editor.tsx

```diff
@@ -0,0 +1,145 @@
+import React, { useEffect, useState, ReactElement } from "react";
+
+type PredefinedQuery = {
+  name: string;
+  sql: string;
+  description: string;
+};
+
+function QueryEditor(props: {
+  onQueryUpdate: (query: string) => void;
+  predefinedQueryOptions?: Array<PredefinedQuery>;
+}) {
+  const [query, setQuery] = useState<string>("");
+  const [queryTemplate, setQueryTemplate] = useState<string>("");
+  const [queryParamValues, setQueryParamValues] = useState<{
+    [key: string]: string;
+  }>({});
+  const [selectedPredefinedQuery, setSelectedPredefinedQuery] = useState<
+    PredefinedQuery | undefined
+  >(undefined);
+
+  const variableRegex = /{{([a-zA-Z0-9_]+)}}/;
+  const textAreaStyle = { width: "100%", height: 140 };
+
+  useEffect(() => {
+    generateQuery();
+  }, [queryTemplate, queryParamValues]);
+
+  useEffect(() => {
+    onQueryTemplateUpdate();
+  }, [queryTemplate]);
+
+  function generateQuery() {
+    let sql = queryTemplate;
+    Object.keys(queryParamValues).forEach((param) => {
+      if (queryParamValues[param]) {
+        sql = sql.split(param).join(queryParamValues[param]);
+      }
+    });
+    setQuery(sql);
+    props.onQueryUpdate(sql);
+  }
+
+  function onQueryTemplateUpdate() {
+    let paramNames = new Set(
```

Comment:
same thing here, you have a function that is taking the query string and returning an array of parameters, that can be tested on its own